File: wallet/src/main/java/com/worldventures/wallet/service/firmware/command/FetchFirmwareUpdateData.java
package com.worldventures.wallet.service.firmware.command;

import com.worldventures.janet.injection.InjectableAction;
import com.worldventures.wallet.domain.entity.FirmwareUpdateData;
import com.worldventures.wallet.service.firmware.FirmwareRepository;

import org.immutables.value.Value;
import org.jetbrains.annotations.Nullable;

import javax.inject.Inject;

import io.techery.janet.Command;
import io.techery.janet.command.annotations.CommandAction;

@CommandAction
public class FetchFirmwareUpdateData extends Command<FetchFirmwareUpdateData.Result> implements InjectableAction {

   @Inject FirmwareRepository firmwareRepository;

   @Override
   protected void run(CommandCallback<Result> callback) throws Throwable {
      final FirmwareUpdateData firmwareUpdateData = firmwareRepository.getFirmwareUpdateData();
      callback.onSuccess(ImmutableResult.builder()
            .isForceUpdateStarted(firmwareUpdateData != null && firmwareUpdateData.isStarted() && firmwareUpdateData.factoryResetRequired())
            .firmwareUpdateData(firmwareUpdateData)
            .build());
   }

   @Value.Immutable
   public static abstract class Result {

      @Deprecated
      @Value.Derived
      public boolean hasUpdate() {
         return isForceUpdateStarted();
      }

      public abstract boolean isForceUpdateStarted();

      @Nullable
      public abstract FirmwareUpdateData firmwareUpdateData();
   }
}

File: wallet/src/main/java/com/worldventures/wallet/service/firmware/command/FetchFirmwareUpdateData.kt
package com.worldventures.wallet.service.firmware.command


import com.worldventures.janet.injection.InjectableAction
import com.worldventures.wallet.domain.entity.FirmwareUpdateData
import com.worldventures.wallet.service.firmware.FirmwareRepository
import io.techery.janet.Command
import io.techery.janet.command.annotations.CommandAction
import javax.inject.Inject

@CommandAction
class FetchFirmwareUpdateData : Command<FetchFirmwareUpdateData.Result>(), InjectableAction {

   @Inject lateinit var firmwareRepository: FirmwareRepository

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<Result>) {
      val firmwareUpdateData = firmwareRepository.firmwareUpdateData
      callback.onSuccess(Result(
            isForceUpdateStarted = firmwareUpdateData != null && firmwareUpdateData.isStarted && firmwareUpdateData.isFactoryResetRequired,
            firmwareUpdateData = firmwareUpdateData))
   }

   data class Result(
      val isForceUpdateStarted: Boolean,
      val firmwareUpdateData: FirmwareUpdateData? = null) {

      @Deprecated("Use isForceUpdateStarted")
      fun hasUpdate() = isForceUpdateStarted
   }
}
